File: wrappers/nodejs/types/vc/core/signer.ts
import {
  _UniversalDIDResolver,
  _VCCoreCredentialSigner,
  Credential,
  Kms,
  UniversalDIDResolver,
  UnsignedCredential,
} from "../../..";
import { contextEnsuredKms } from "../../utils";

/**
 * An async low-level `SignCredential` API.
 *
 * Carries only what the signing step needs — a {@link Kms} for key access and a
 * DID resolver for LDP proof assembly. Use when the prepare/sign split is
 * driven from application code (e.g. a remote signing service or HSM-backed
 * adapter) and a full {@link VCCoreIssuer} is not desired.
 *
 * This class is a hand-written wrapper over the raw napi `_VCCoreCredentialSigner`:
 * - it threads the {@link Kms} through {@link contextEnsuredKms} so the napi
 *   `ThreadsafeFunction` machinery preserves `this` on callbacks;
 * - it accepts the friendly {@link UniversalDIDResolver} (also a TS wrapper) and
 *   unwraps its inner napi resolver.
 *
 * The resulting API matches the wasm-side `VCCoreCredentialSigner`, so the same
 * application code runs against either binding.
 */
export class VCCoreCredentialSigner {
  private readonly inner: _VCCoreCredentialSigner;

  constructor(kms: Kms, didResolver: UniversalDIDResolver) {
    this.inner = new _VCCoreCredentialSigner(contextEnsuredKms(kms), didResolver.inner);
  }

  /**
   * Sign an unsigned credential and return the finished {@link Credential}.
   *
   * The `unsigned` argument is the externally-tagged JSON serialisation of the
   * SDK's {@link UnsignedCredential} enum — `{ "SdJwt": { ... } }` or
   * `{ "Ldp": { ... } }`.
   */
  async signCredential(unsigned: UnsignedCredential): Promise<Credential> {
    return await this.inner.signCredential(unsigned);
  }
}

File: wrappers/uniffi/src/vc/core_api/signer.rs
use crate::common::{Error, JsonValue, Result};
use crate::did::universal_resolver::UniversalDIDResolver;
use crate::key_handle::WrappedKeyHandle;
use crate::kms::{Kms, WrappedKms};
use crate::vc::Credential;
use agent_sdk::vc::core::{
    CredentialSigner as CoreCredentialSigner, SignCredential, UnsignedCredential,
};
use std::sync::Arc;

/// An async low-level `SignCredential` API.
///
/// Carries only what the signing step needs — a `Kms` for key access and a
/// DID resolver for LDP proof assembly. Use when the prepare/sign split is
/// driven from application code and a full issuer service is not desired.
#[derive(uniffi::Object)]
pub struct VCCoreCredentialSigner(Box<dyn SignCredential>);

#[uniffi::export(async_runtime = "tokio")]
impl VCCoreCredentialSigner {
    /// Create a new credential signer.
    #[uniffi::constructor]
    pub fn new(kms: Arc<dyn Kms>, did_resolver: Arc<UniversalDIDResolver>) -> Self {
        let signer: CoreCredentialSigner<WrappedKeyHandle, WrappedKms> =
            CoreCredentialSigner::new(WrappedKms::new(kms), did_resolver.inner());
        Self(Box::new(signer))
    }

    /// Sign an unsigned credential and return the finished `Credential`.
    ///
    /// The `unsigned_credential` argument is the externally-tagged JSON shape of
    /// the SDK's `UnsignedCredential` enum — `{ "SdJwt": { ... } }` or
    /// `{ "Ldp": { ... } }`.
    pub async fn sign_credential(&self, unsigned_credential: JsonValue) -> Result<Credential> {
        let unsigned: UnsignedCredential = serde_json::from_value(unsigned_credential)
            .map_err(|err| Error::OID4VCIInternal(format!("invalid UnsignedCredential: {err}")))?;
        self.0
            .sign_credential(unsigned)
            .await
            .map_err(|err| Error::OID4VCIInternal(format!("{err}")))
    }
}
